sdk/webpubsub/azure-messaging-webpubsub-client/src/main/java/com/azure/messaging/webpubsub/client/models/WebPubSubProtocol.java

```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.messaging.webpubsub.client.models;
+
+/**
+ * The superclass of protocol for WebPubSub client.
+ */
+public abstract class WebPubSubProtocol {
```

Comment:
The alternative I can think of is an extensible enum `WebPubSubProtocol`, with enum and additional value be provider name e.g. "com.azure.messaging.webpubsub.protocol.WebPubSubJsonReliableProtocolProvider".

Runtime, Builder would use [ServiceLoader](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html) to look up available protocol in runtime, pick up the one with matching name. Throw exception when not found.

One benefit maybe that, we do not need to expose the `WebPubSubJsonReliableProtocol` class to user. It could be in implementation package with a `open to`.

Reply:
We could provide an overload with the extensible enum, like charset and codex.

The subprotocol likely not configurable. If it does, user can fallback to the WebPubSubProtocol subclass for additional configuration.

Reply:
Their .NET design is to provide a e.g. WebPubSub.Protocol.ProtoBuf package, which contains some subclass of `WebPubSubProtocol`

If we use enum (probably just `Enum`), I don't think we can subclass the Enum?